cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/join/Join.java

```diff
@@ -0,0 +1,108 @@
+ */
+
+package co.cask.cdap.etl.batch.join;
+
+import co.cask.cdap.etl.api.Emitter;
+import co.cask.cdap.etl.api.JoinConfig;
+import co.cask.cdap.etl.api.JoinElement;
+import co.cask.cdap.etl.api.JoinResult;
+import co.cask.cdap.etl.api.Joiner;
+import com.google.common.collect.Iterables;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Performs join
+ * @param <JOIN_KEY> type of join key
+ * @param <INPUT_RECORD> type of input record
+ * @param <OUT> type of output of mapreduce
+ */
+
+public class Join<JOIN_KEY, INPUT_RECORD, OUT> {
+  private Joiner<JOIN_KEY, INPUT_RECORD, OUT> joiner;
+  private JOIN_KEY joinKey;
+  private Iterator<JoinElement<INPUT_RECORD>> iterator;
+  private Emitter<OUT> emitter;
+  private final int numOfInputs;
+
+  public Join(Joiner<JOIN_KEY, INPUT_RECORD, OUT> joiner, JOIN_KEY joinKey,
+              Iterator<JoinElement<INPUT_RECORD>> iterator, int numOfInputs, Emitter<OUT> emitter) throws Exception {
+    this.joiner = joiner;
+    this.joinKey = joinKey;
+    this.iterator = iterator;
+    this.numOfInputs = numOfInputs;
+    this.emitter = emitter;
+  }
+
+  public void joinRecords() throws Exception {
+    Map<String, List<JoinElement<INPUT_RECORD>>> perStageJoinElements = getPerStageJoinElements();
+    JoinConfig joinConfig = joiner.getJoinConfig();
+    Iterable<String> requiredInputs = joinConfig.getRequiredInputs();
+    if (Iterables.size(requiredInputs) == numOfInputs) {
+      innerJoin(perStageJoinElements);
+    }
+  }
+
+  private Map<String, List<JoinElement<INPUT_RECORD>>> getPerStageJoinElements() {
+    Map<String, List<JoinElement<INPUT_RECORD>>> perStageJoinElements = new HashMap<>();
+    while (iterator.hasNext()) {
+      JoinElement<INPUT_RECORD> joinElement = iterator.next();
+      String stageName = joinElement.getStageName();
+      if (perStageJoinElements.get(stageName) == null) {
+        perStageJoinElements.put(stageName, new ArrayList<JoinElement<INPUT_RECORD>>());
+      }
+      perStageJoinElements.get(stageName).add(joinElement);
+    }
+    return perStageJoinElements;
+  }
+
+  private void innerJoin(Map<String, List<JoinElement<INPUT_RECORD>>> perStageJoinElements) throws Exception {
+    // As we get intersection of records from n stages in inner join, if the number of stages we got after reduce
+    // are not same as number of stages we are joining, then there is nothing to emit.
+    if (perStageJoinElements.size() != numOfInputs) {
+      return;
+    }
+
+    List<List<JoinElement<INPUT_RECORD>>> list = new ArrayList<>(perStageJoinElements.values());
+    // cross product of each stage record
+    ArrayList<JoinElement<INPUT_RECORD>> joinElements = new ArrayList<>();
+    ArrayList<JoinResult<INPUT_RECORD>> joinResults = new ArrayList<>();
+    getCartesianProduct(list, 0, numOfInputs, joinElements, joinResults);
+    joiner.emit(joinKey, joinResults, emitter);
```

Comment:
I thought the app was supposed to be responsible for joining everything, and the Joiner plugin gets joined rows?  This looks like its passing the entire cartesian product to the Joiner plugin